src/Analyzers/CSharp/Analyzers/ConstantInterpolatedString/CSharpConstantInterpolatedStringAnalyzer.cs

```diff
@@ -0,0 +1,88 @@
+    internal sealed class CSharpConstantInterpolatedStringAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public CSharpConstantInterpolatedStringAnalyzer() : base(
+            IDEDiagnosticIds.UseConstantInterpolatedStringDiagnosticId,
+            EnforceOnBuildValues.UseConstantInterpolatedString,
+            option: null,
+            new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_constant_interpolated_string), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!((CSharpCompilation)context.Compilation).LanguageVersion.HasConstantInterpolatedStrings())
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(context =>
+                {
+                    var operation = (IBinaryOperation)context.Operation;
+                    if (!ShouldAnalyze(operation))
+                    {
+                        return;
+                    }
+
+                    if (!AllOperandsAreStringLiterals(operation))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, operation.Syntax.GetLocation()));
+                    }
+                }, OperationKind.Binary);
+            });
+        }
+
+        private static bool AllOperandsAreStringLiterals(IBinaryOperation? operation)
+        {
+            while (operation is not null)
+            {
+                if (operation.LeftOperand.Kind != OperationKind.Literal)
+                {
+                    return false;
+                }
+
+                if (operation.RightOperand.Kind == OperationKind.Literal)
+                {
+                    return true;
+                }
+
+                operation = operation.RightOperand as IBinaryOperation;
+            }
+
+            return false;
+        }
+
+        private static bool ShouldAnalyze(IBinaryOperation operation)
+        {
+            // Avoid analyzing user-defined operators.
+            if (operation.OperatorMethod is not null ||
+                operation.OperatorKind != BinaryOperatorKind.Add ||
+                // Avoid nested diagnostics. We will report on the parent if needed.
+                operation.Parent?.Kind == OperationKind.Binary ||
+                operation.ConstantValue.Value is not string)
+            {
+                return false;
+            }
+
+            // A heuristic for constant-only context.
+            return operation.Syntax.Parent.IsKind(SyntaxKind.AttributeArgument) ||
+                operation.Syntax.FirstAncestorOrSelf<LocalDeclarationStatementSyntax>()?.IsConst == true;
```

Comment:
Constant fields?

Reply:
Will do

Reply:
I removed this check altogether. It doesn't harm if we produced a diagnostic in non-constant context. User can still use constant interpolations.